src/main/java/at/ac/tuwien/kr/alpha/grounder/instantiation/LiteralInstantiationResult.java

```diff
@@ -0,0 +1,135 @@
+	 * types STOP_BINDING, PUSH_BACK and MAYBE_PUSH_BACK.
+	 */
+	private final Optional<List<ImmutablePair<Substitution, AssignmentStatus>>> substitutions;
+
+	private LiteralInstantiationResult(Type type, Optional<List<ImmutablePair<Substitution, AssignmentStatus>>> substitutions) {
+		this.type = type;
+		this.substitutions = substitutions;
+	}
+
+	public static LiteralInstantiationResult continueBinding(List<ImmutablePair<Substitution, AssignmentStatus>> substitutions) {
+		return new LiteralInstantiationResult(Type.CONTINUE, Optional.of(substitutions));
+	}
+
+	//@formatter:off
+	public static LiteralInstantiationResult continueBindingWithTrueSubstitutions(List<Substitution> substitutions) {
+		List<ImmutablePair<Substitution, AssignmentStatus>> substitutionsWithAssignment = substitutions.stream()
+				.map((substitution) -> new ImmutablePair<>(substitution, AssignmentStatus.TRUE))
+				.collect(Collectors.toList());
+		return new LiteralInstantiationResult(Type.CONTINUE, Optional.of(substitutionsWithAssignment));
+	}
+	//@formatter:on
+
+	public static LiteralInstantiationResult continueBinding(Substitution substitution, AssignmentStatus assignmentStatus) {
+		return new LiteralInstantiationResult(Type.CONTINUE, Optional.of(Collections.singletonList(new ImmutablePair<>(substitution, assignmentStatus))));
+	}
+
+	public static LiteralInstantiationResult stopBinding() {
+		return new LiteralInstantiationResult(Type.STOP_BINDING, Optional.empty());
+	}
+
+	public static LiteralInstantiationResult pushBack() {
+		return new LiteralInstantiationResult(Type.PUSH_BACK, Optional.empty());
+	}
+
+	public static LiteralInstantiationResult maybePushBack() {
+		return new LiteralInstantiationResult(Type.MAYBE_PUSH_BACK, Optional.empty());
+	}
+
+	public Type getType() {
+		return this.type;
+	}
+
+	// unwrap the optional, if empty, exception will be thrown here
+	public List<ImmutablePair<Substitution, AssignmentStatus>> getSubstitutions() {
+		return this.substitutions.get();
+	}
+
+	/**
+	 * Result type. Used by {@link Grounder}s to determine how to proceed grounding a rule after instantiating the literal that yielded the
+	 * respective instantiation result.
+	 * 
+	 * Copyright (c) 2020, the Alpha Team.
+	 */
+	public static enum Type {
+
+		/**
+		 * Grounder should stop working on the rule, no valid substitutions exist.
+		 */
+		STOP_BINDING,
+
+		/**
+		 * Grounder should continue with the next literal.
+		 */
+		CONTINUE,
+
+		/**
+		 * Literal instantiation yielded no ground instances, but depending on the specific workflow of the grounder, proceeding with another
+		 * literal might yet make sense. Grounder should decide whether to stop binding or push the literal back in the overall grounding order.
+		 */
+		MAYBE_PUSH_BACK,
+
+		/**
+		 * Currently no ground instances, but proceeding with another literal might make sense, push the literal back in the overall grounding
```

Comment:
Yes, in the sense that in both cases we do not *know* whether pushing back will really cause the literal to have substitutions then, but there are grounds to assume it might be so. The difference between the two is that for some cases, e.g.  `FixedInterpretationLiteral`s we will always try to push back, while in some cases whether to push back is dependent on the permissiveness of the used heuristic and therefore in `NaiveGrounder`s area of responsibility.